### PR TITLE
Augment Intro to Tests topic

### PR DESCRIPTION
**Pull request by beccaelenzil**
Main changes/additions:
- Outlined different types of test cases in the Intro to Automated Tests lesson
- Added a lesson on TDD
- Updated TDD Exercise (black jack) and added thinking about test cases to the problem set (Part 1), and Writing Tests (Part 2 and 3) to an activity: https://github.com/AdaGold/tdd-exercise/pull/4
- Added an instructor doc with documentation of TDD Ping Pong, a suggested livecode: https://github.com/AdaGold/tdd-ping-pong/pull/1 
- Removed last problem set question that is about testing for handling exceptions, and copied this to the handling exceptions topic to potentially be used later.

Heavily pulled from this lesson: https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/main/00-programming-fundamentals/intro-to-automated-tests.md

The additions do not touch on testing more broadly than unit tests. A callout mentioned that integration tests will be introduced with projects. My reasoning was the Topic/lessons seemed big enough as they are for a single day/topic, but am open to thoughts on introducing testing more broadly.

### CODE REVIEW

File: intro-to-tests/automated-tests.md [hunk 1]
```diff
@@ -36,13 +36,22 @@ Each test could follow this generic pattern:
...
+<!-- available callout types: info, success, warning, danger, secondary  -->
+
 ![Part of an astrology birth chart](../assets/intro-to-tests_automated-tests_astrology.jpg)  
 [(source)](https://pixabay.com/photos/astrology-divination-chart-993127/)
 
 Addison is writing software about astrology. They are working on a feature that, when a user provides a date, time, and location of their birth, will compute their astrological birth chart. They need to know if they've implemented the logic correctly; does the feature accurately compute the astrological birth chart? They can test their feature by finding the following information:
 
 1. Get a sample of at least one correct example: an input and and output
-   - This means, finding and using a real date, time, and location, and its real astrological birth chart
+   - This means,∂ finding and using a real date, time, and location, and its real astrological birth chart
```
Comment (CheezItMan) [inline intro-to-tests/automated-tests.md:54]: Extra character?

```suggestion
   - This means, finding and using a real date, time, and location, and its real astrological birth chart
```

File: intro-to-tests/automated-tests.md [hunk 2]
```diff
@@ -36,13 +36,22 @@ Each test could follow this generic pattern:
 
 ### Example: Astrology Software
 
+### !callout-info
+
+## Astrological Birth Chart
+An Astrological Birth Chart is a visual map of what the sky looked like when you were born. It is determined using your exact birthdate and birthplace. It includes three signs, your **sun sign** (typically refered to as your Zodiac sign), your **moon sign**, and your **rising sign**. 
+[source](https://www.shape.com/lifestyle/mind-and-body/sun-moon-rising-sign-meaning)
```
Comment (CheezItMan) [inline intro-to-tests/automated-tests.md:43]: Birth charts are pretty complicated and lots of people (like me) don't know anything about them.

File: intro-to-tests/automated-tests.md [hunk 3]
```diff
@@ -66,21 +75,46 @@ Because there are a lot of different kinds of code, there are a lot of different
 
 ### Test Cases and Scenarios
 
-Testing a single function could have many unit tests to test it. Why? Based on the context, a single function could potentially have different arguments that give back different return values. Different scenarios in development can imply different arguments, which would produce different logic. All scenarios and possible outcomes of a function should be tested, as much as reasonably possible.
+Testing a single function will generally have many unit tests to test it. Why? Based on the context, a single function could potentially have different arguments that give back different return values. Different scenarios in development can imply different arguments, which would produce different logic. All scenarios and possible outcomes of a function should be tested, as much as reasonably possible.
```
Comment (CheezItMan) [inline intro-to-tests/automated-tests.md:78]:
```suggestion
Testing a single function will generally have many unit tests to ensure the code works correctly. Why? Based on the context, a single function could potentially have different arguments that give back different return values. Different scenarios in development can imply different arguments, which would produce different logic. All scenarios and possible outcomes of a function should be tested, as thoroughly as possible.
```

File: intro-to-tests/problem-set-intro-to-pytest.md [hunk 4]
```diff
@@ -1,10 +1,37 @@
...
+* id: 6cc1f1e5-b54a-4d89-900f-2d18f8db240d
+* title: Identifying Tests Cases
+<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
+<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->
+
+##### !question
+
+Complete Part 1 outlined in the readme of the [tdd-exercise repository](https://github.com/AdaGold/tdd-exercise).
+
+Record your answers below.
```
Comment (CheezItMan) [inline intro-to-tests/problem-set-intro-to-pytest.md:21]: Just a suggestion to make it clear they're copying and pasting their test cases here. 

```suggestion
Record your test cases below.
```

File: intro-to-tests/test-driven-development.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+We should run our **entire** automated unit test suite often. When our requirements change, we can feel secure in changing our code because we have some _tests_ to let us know if we've broken something.
+
+__Tests are transient.__ As we work on a project, our understanding of the problems at hand will change. As they do, your tests will change.
+
+Keep in mind the balance between evolving requirements, evolving tests, and tests that should stay the same. This balance will change with every task.
+
+### !callout-info
+
+## Integration Tests
+The tests we write as part of the TDD programming workflow described here are are *Unit Tests*, tests that to check a single function. There are other types of test. One other type of test we'll see in this curriculum are *Integration Tests*, tests that determine if software works when multiple parts are connected together. We will learn more about Integration Tests with our first project.
```
Comment (CheezItMan) [inline intro-to-tests/test-driven-development.md:56]: There might be a case to put in [this graphic](https://3kllhk1ibq34qk6sp3bhtox1-wpengine.netdna-ssl.com/wp-content/uploads/03-Piramides-03-03-1400x531.png) or something like it somewhere illustrating the types of tests.

**Review by CheezItMan — APPROVED** [review]
Overall I think this is a good addition.  In the future maybe some kind of reading on types of testing might be good, but it's not directly useful for students.

File: intro-to-tests/automated-tests.md [hunk 6]
```diff
@@ -66,21 +75,46 @@ Because there are a lot of different kinds of code, there are a lot of different
...
+| 1. Do we get an accurate birth chart if we pass in a date, location, and time? | `"January 1, 1980"`, `"Seattle"`, `"12:00"` | `{"Sun": "Capricorn", ...}` |
+| 2. What happens if the function receives `None` for time?                      | `"January 1, 1980"`, `"Seattle"`, `None`    | `None`                      |
+| 3. What happens if the `date` argument comes as a different format?            | `"1980-01-01"`, `"Seattle"`, `"12:00"`      | `{"Sun": "Capricorn", ...}` |
+| 4. What happens if the function receives an invalid place for `location`?       | `"January 1, 1980"`, `"xyz"`, `"12:00"` | should raise an exception   |
+
+To ensure that our test cases cover all scenarios, it can be helpful to categorize tests using the following categories:
+1. Positive Nominal
+1. Negative Nominal
+1. Positive Edge
+1. Negative Edge
```
Comment (alope107) [inline intro-to-tests/automated-tests.md:93]: Up to your judgement, but I think we don't need to emphasize the positive and negative here; separation into nominal and edge cases feels sufficient.

File: intro-to-tests/automated-tests.md [hunk 7]
```diff
@@ -66,21 +75,46 @@ Because there are a lot of different kinds of code, there are a lot of different
...
+
+#### Nominal Test Cases
+
+A nominal test case is a type of test case that describes a piece of core functionality needed for the success of this method. This is the test case that verifies that the method does its primary responsibility. Nominal test cases are the obvious test cases.
+
+In the case of Addison's astrological software, the first test case listed above is a nominal **positive nominal** test case. It checks that given the correct, expected input (the `date`, `location`, and `time`), the functions returns the correct output (a dictionary with the sun, rising, and moon signs). 
+
+A **negative nominal** test case is a type of test case that describes the most obvious, most typical way that a function handles unexpected input. Again, consider our test cases outlined above, the second test case falls under this category. If we pass in `None` for time, the function
+does not have enough information to determine the astrological birth chart, and as such, it returns `None`.
```
Comment (alope107) [inline intro-to-tests/automated-tests.md:102]: I've never heard of a negative nominal case, and I'm not seeing it coming up in Google searches either. To me, `None` feels like an edge case, and we can remove discussion of negative nominal entirely.

File: intro-to-tests/automated-tests.md [hunk 8]
```diff
@@ -66,21 +75,46 @@ Because there are a lot of different kinds of code, there are a lot of different
...
+
+To ensure that our test cases cover all scenarios, it can be helpful to categorize tests using the following categories:
+1. Positive Nominal
+1. Negative Nominal
+1. Positive Edge
+1. Negative Edge
+
+#### Nominal Test Cases
+
+A nominal test case is a type of test case that describes a piece of core functionality needed for the success of this method. This is the test case that verifies that the method does its primary responsibility. Nominal test cases are the obvious test cases.
```
Comment (alope107) [inline intro-to-tests/automated-tests.md:97]: Consider adding that nominal test cases are often referred to as "happy path" test cases.

File: intro-to-tests/automated-tests.md [hunk 9]
```diff
@@ -66,21 +75,46 @@ Because there are a lot of different kinds of code, there are a lot of different
...
+does not have enough information to determine the astrological birth chart, and as such, it returns `None`.
+
+#### Edge Test Cases
+
+An edge test case is a type of test case that verifies that the function can work successfully, even given non-obvious context. This is the test case that verifies that the function works, even with very unexpected context.
+
+A **positive edge** test case is a test case that describes a set of inputs and expectations that are on the limits of the method's most obvious, most typical way of working successfully. In the test cases listed above, the third test case is a positive edge case. The function returns the correct birth chart even when the input is in an unexpected format.
+
+A **negative edge** test case is a test case that tests that this function can handle the most non-obvious, most atypical unexpected input gracefully. The fourth example above is a negative edge case.The function receives an invalid place for `location` and raises an exception. We will learn more about raising, handling, and testing exceptions in [Exception Handling](../exception-handling/raising-and-handling-exceptions.md)
```
Comment (alope107) [inline intro-to-tests/automated-tests.md:110]:
```suggestion
A **negative edge** test case is a test case that tests that this function can handle the most non-obvious, most atypical unexpected input gracefully. The fourth example above is a negative edge case. The function receives an invalid place for `location` and raises an exception. We will learn more about raising, handling, and testing exceptions in [Exception Handling](../exception-handling/raising-and-handling-exceptions.md)
```

File: intro-to-tests/automated-tests.md [hunk 10]
```diff
@@ -140,7 +177,7 @@ Why are automated tests beneficial?
 ##### !end-answer
 
 ##### !explanation 
-Tests and these benefits are reasons to follow a practice called test-driven development. If you'd like to look that up, follow your curiosity~
+Tests and these benefits are reasons to follow a practice called test-driven development.~
```
Comment (alope107) [inline intro-to-tests/automated-tests.md:180]:
```suggestion
Tests and these benefits are reasons to follow a practice called test-driven development.
```

**Review by alope107 — COMMENTED** [review]
Looking good! Added a few in-line comments.

File: exception-handling/extra-problem-set-question.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
+# Question moved from Intro to Tests topic, may be useful in Exception Handling topic
```
[the remark this answers is not in the document]
Comment (beccaelenzil) [reply exception-handling/extra-problem-set-question.md:1]: thanks! it's just there for curriculum devs as a resource, but certainly open to better ways to keep track of these sorts of resources

File: intro-to-oop-relationships/problem-set-oop-relationships.md [hunk 12]
```diff
@@ -204,6 +204,7 @@ print(bear.make_noise())
 ##### !end-options
 
 ##### !answer
+*
```
[the remark this answers is not in the document]
Comment (beccaelenzil) [reply intro-to-oop-relationships/problem-set-oop-relationships.md:207]: This was intentional! There was a bug where it wouldn't publish on Learn because the answer was missing a `*`. Note to future selves: answers for checkbox questions need a `*`.

File: intro-to-tests/automated-tests.md (hunk 2 above)
Comment (beccaelenzil) [reply intro-to-tests/automated-tests.md:43]: I agree, me too. That's why I added this callout to provide some context. Do you think more context is needed to ensure this example is understandable?

**Review by beccaelenzil — COMMENTED** [review]
Thank you for your feedback! I incorporated all of it.
